Main/Game/Game/Game.cpp

```diff
@@ -0,0 +1,334 @@
+                    if (dx || dy) allTilesInPlace = false;
+                }
+            }
+            if (allTilesInPlace) isReshuffling = false;
+        }
+
+        if (movesLeft <= 0 && !gameOver && !gameWon) {
+            gameOver = true;
+        }
+        if (taskManager.isTaskCompleted() && !gameOver && !gameWon) {
+            gameWon = true;
+        }
+
+        app.clear();
+        if (gameOver) {
+            app.draw(gameOverBackground);
+        } else if (gameWon) {
+            app.draw(gameWonBackground);
+        } else {
+            app.draw(background);
+            app.draw(movesText);
+            taskManager.draw(app, font, sf::Vector2f(app.getSize().x - 400, 10));
+            for (int row = 0; row < GAME_SIZE; row++) {
+                for (int col = 0; col < GAME_SIZE; col++) {
+                    Tile* p = grid[row][col];
+                    p->sprite.setColor(sf::Color(255, 255, 255, p->alpha));
+                    p->sprite.setPosition(p->x + offset.x, p->y + offset.y);
+                    app.draw(p->sprite);
+                }
+            }
+            if (!gameOver && !gameWon) {
+                app.draw(shuffleButton);
+            }
+        }
+        app.display();
+    }
+}
+
+void Match3Game::restartGame() {
+    srand(time(0));
+    for (int row = 0; row < GAME_SIZE; row++) {
+        for (int col = 0; col < GAME_SIZE; col++) {
+            delete grid[row][col];
+            const unsigned int TileType = rand() % 4;
+            switch (TileType) {
+                case 0:
+                    grid[row][col] = new RedTile(row, col);
+                    break;
+                case 1:
+                    grid[row][col] = new BlueTile(row, col);
+                    break;
+                case 2:
+                    grid[row][col] = new GreenTile(row, col);
+                    break;
+                case 3:
+                    grid[row][col] = new YellowTile(row, col);
+                    break;
+            }
+        }
+    }
+    movesLeft = 30;
+    taskManager.setTask(TileType::Red, 3, "images/NewNodes/Stage3/FrogP.png");
+}
+
+void Match3Game::reshuffleGrid() {
+    std::vector<Tile*> tiles;
+    for (int row = 0; row < GAME_SIZE; row++) {
+        for (int col = 0; col < GAME_SIZE; col++) {
+            tiles.push_back(grid[row][col]);
+        }
+    }
+    std::shuffle(tiles.begin(), tiles.end(), std::default_random_engine(std::time(0)));
+    for (int i = 0; i < tiles.size(); i++) {
+        int row = i / GAME_SIZE;
+        int col = i % GAME_SIZE;
+        tiles[i]->row = row;
+        tiles[i]->col = col;
+        tiles[i]->y = -TILE_SIZE * (GAME_SIZE - row);
+        grid[row][col] = tiles[i];
+    }
+
+    for (int row = 0; row < GAME_SIZE; row++) {
+        for (int col = 0; col < GAME_SIZE; col++) {
+            const unsigned int TileType = rand() % 4;
+            switch (TileType) {
+                case 0:
+                    delete grid[row][col];
```

Comment:
no need to call `delete grid[row][col]` in every case. 
Just call it once before the switch